# Football Manager 2020 Original Dataset

This dataset contains a collection of players (no newgen) and their attributes, such as age, position, club, nationality, value, wage, all player attribute components, etc.

### Import Library

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### Load Dataset

In [2]:
df = pd.read_csv('datafm20.csv')
df.head()

Unnamed: 0               Name                Position       Club  \
0           0       Lionel Messi         AM (RC), ST (C)  Barcelona   
1           1  Cristiano Ronaldo         AM (RL), ST (C)   Juventus   
2           2     Kylian Mbappé         AM (RL), ST (C)   Paris SG   
3           3       Manuel Neuer                      GK  FC Bayern   
4           4             Neymar  M (L), AM (LC), ST (C)   Paris SG   

                 Division                       Based Nation  Height  Weight  \
0  Spanish First Division      Spain (First Division)    ARG     170      72   
1         Italian Serie A             Italy (Serie A)    POR     185      83   
2       Ligue 1 Conforama  France (Ligue 1 Conforama)    FRA     178      73   
3              Bundesliga        Germany (Bundesliga)    GER     192      90   
4       Ligue 1 Conforama  France (Ligue 1 Conforama)    BRA     175      68   

   Age  ... Cmp Com Cmd  Bra  Bal  Ant  Agi  Agg  Aer  Acc  
0   32  ...  18   2   2   10   19   19   19    7    2   18  
1   34  ...  14   4   1   16   14   18   13    6    2   15  
2   20  ...  18   2   1   12   14   17   16    6    4   20  
3   33  ...  20  12  17   14   14   19   15    8   14   13  
4   27  ...  17   3   1   14   14   16   18   10    4   17  

[5 rows x 64 columns]

#### Column Description

Profile
* Name: Name of player
* Position: All possible position of Player
* Club: Club of player
* Division: Division of player's club
* Based: Nation Based of player
* Nation: Nationality of player
* Age: Age of player
* Preferred Foot: Preferred Foot of player
* Best Pos: Best position of player
* Best Role: Best role of player in position
* Value: Current value of player (in euro)
* Wage: Wage of player in club (in euro, per week)
* CA: Current Ability of player (hidden attribute in FM20)
* PA: Potential Ability of player (hidden attribute in FM20)

Mental Attributes
* Wor: Work Rate; Player's willingness to work to his full capacity
* Vis: Vision; Player's ability to see a potential opening and spot an opportunity another player may not have seen
* Tea: Teamwork; Player can follow tactical instructions whilst working for and alongside his team-mates
* Pos: Positioning; Player's ability to read a defensive situation and position themselves accordingly
* OtB: Off the Ball; Player's ability to move when not in possession of the ball, making themselves available to receive a pass in dangerous position
* Ldr: Leadership; Player's ability to influence player around them on the pitch
* Fla: Flair; Player's natural talent for the creative and unpredictable
* Det: Determination; Player's commitment to succeed and do his very best on and off the pitch
* Dec: Decision; Player's ability to make the correct choice both with and without the ball
* Cnt: Concentration; Player's mental focus and attention to detail on a event a event basis
* Cmp: Composure; Player's steadiness of mind and ability to make intelligent decisions with and without the ball
* Bra: Bravery; How committed player is, often putting themselves into risky situations which may cause injury
* Ant: Anticipation; Player's ability to predict and react to events going on around them
* Agg: Aggression; Player willingness to get stuck in, perharps at expense of giving away more fouls

Technical Attributes
* Tec: Technique; The aesthetic quality of a player's technical game
* Tck: Tackling; Player's ability to win the ball cleanly without conceding foul in such situations
* Pen: Penalty Taking; Player;s ability from the penalty spot
* Pas: Passing; Player's ability to successfully find a team-mate with the ball
* 1v1: One on Ones; Goalkeeper's ability to do well when faced with an opponent in a one on one situtation
* Mar: Marking; Player's ability to stick close to his direct opposition in defensive situations
* L Th: Long Throws; Player's ability to throw the balllong, often in attacking situations
* Lon: Long Shots; Player's prowess when shooting from outsine penalty area
* Hea: Heading; Player's ability to head the ball
* Fre: Freekick Taking; Player's ability to strike a dead ball, either when having a shot or putting the ball into dangerous areas from deep or out wide
* Fir: First Touch; Player's ability to control the ball immediately as it is passed into feet
* Fin: Finishing; Player's ability to putt the ball in the back of the net when presented with a chance
* Dri: Dribbling; Player's ability to run with the ball and manipulate it under close control
* Cro: Crossing; Player's ability to cross the ball accurately from wide areas
* Cor: Corner; Player's ability to accurately take a corner

Physical Attributes
* Str: Strength; Player;s ability to exert his physical force on an opponent to his benefit
* Sta: Stamina; Player;s ability to endure high-level physical activity for extended periods of time
* Pac: Pace; Player's top speed both on and off the ball
* Nat: Natural Fitness; How well a player stays fit when injured or not in training
* Jum: Jumping Reach; The highest point that a player can reach with his head
* Bal: Balance; How well a player can stay on his feet, both on and off the ball
* Agi: Agility; How well player can start, stop and move in different directions at varying levels of speed both on and off the ball
* Aer: Aerial Reach; Player's physical ability to challange in aerial situations
* Acc: Acceleration; How quickly a player can get to top of speed from standing start
* Height: Height of player (in cm)
* Weight: Weight of player (in kg)

Goalkeeping Attributes
* Thr: Throwing; Goalkeeper's ability to accurately distribute the ball from his hands
* TRO: Rushing Out (Tendency); Goalkeeper's tendency to come off his line toreact to through ball and crosses
* Ref: Reflexes; Goalkeeper's ability to react to unpredictable events with a high success rate
* Pun: Punching; Goalkeeper's inclination to punch the ball clear in situations where he could perharps attempt to catch the ball
* Kic: Kicking; The distance goalkeeper can reach with a kick both from his hands and on the ground
* Han: Handling; Goalkeeper's ability to hold onto the ball when making a save
* Ecc: Eccentricity; Goalkeeper's tendency to do the unexpected with or without the ball
* Com: Communication; Goalkeeper's ability to communicate with his defensive line and organise the defensive side of the team
* Cmd: Command of Area; Goalkeeper's ability to instinctively take charge of his penalty area by coming crosses and commanding his defensive line


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144750 entries, 0 to 144749
Data columns (total 64 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   Unnamed: 0      144750 non-null  int64 
 1   Name            144750 non-null  object
 2   Position        144750 non-null  object
 3   Club            144750 non-null  object
 4   Division        144750 non-null  object
 5   Based           144750 non-null  object
 6   Nation          144750 non-null  object
 7   Height          144750 non-null  int64 
 8   Weight          144750 non-null  int64 
 9   Age             144750 non-null  int64 
 10  Preferred Foot  144750 non-null  object
 11  Best Pos        144750 non-null  object
 12  Best Role       144750 non-null  object
 13  Value           144750 non-null  int64 
 14  Wage            144750 non-null  int64 
 15  CA              144750 non-null  int64 
 16  PA              144750 non-null  int64 
 17  Wor             144750 non-nu

#### Observation:

* There are 144750 rows and 64 columns in dataset
* There is one unused column, probably index column for previous dataset format. We need to drop this column!
* Surprisingly there is no null data!
* Each column has been in a proper data type
* Since memory usage for this dataset is 70.7+ MB, let's try to decrease the bytesize to fasten the processing!
* We need to rename some columns to remove the whitespace!
* We need to collect columns based on data type and attribute categories for further analysis!
* We need to check duplicated values

### Central Tendency Measures 

In [4]:
df.describe()

Unnamed: 0         Height         Weight            Age  \
count  144750.00000  144750.000000  144750.000000  144750.000000   
mean    72374.50000     180.352339      72.970515      24.046266   
std     41785.87007       7.056189       7.339469       5.509446   
min         0.00000     150.000000      55.000000      14.000000   
25%     36187.25000     175.000000      68.000000      20.000000   
50%     72374.50000     180.000000      73.000000      23.000000   
75%    108561.75000     185.000000      78.000000      28.000000   
max    144749.00000     208.000000     120.000000      53.000000   

              Value          Wage             CA             PA  \
count  1.447500e+05  1.447500e+05  144750.000000  144750.000000   
mean   3.503583e+05  1.605965e+03      75.816214      94.486017   
std    2.392352e+06  1.037654e+04      24.913951      25.052014   
min    0.000000e+00  0.000000e+00       1.000000       2.000000   
25%    1.700000e+02  0.000000e+00      60.000000      80.000000   
50%    9.000000e+03  1.600000e+02      76.000000      96.000000   
75%    6.500000e+04  6.000000e+02      92.000000     111.000000   
max    9.100000e+07  1.200000e+06     195.000000     200.000000   

                 Wor            Vis  ...            Cmp            Com  \
count  144750.000000  144750.000000  ...  144750.000000  144750.000000   
mean        9.609603       7.854418  ...       8.299731       2.790225   
std         3.001689       3.202000  ...       2.831488       2.342749   
min         1.000000       1.000000  ...       1.000000       1.000000   
25%         7.000000       6.000000  ...       6.000000       1.000000   
50%        10.000000       8.000000  ...       8.000000       2.000000   
75%        12.000000      10.000000  ...      10.000000       3.000000   
max        20.000000      20.000000  ...      20.000000      20.000000   

                 Cmd            Bra            Bal            Ant  \
count  144750.000000  144750.000000  144750.000000  144750.000000   
mean        2.852028       8.683095       8.263157       9.033022   
std         2.481378       3.411490       3.012837       2.722840   
min         1.000000       1.000000       1.000000       1.000000   
25%         1.000000       6.000000       6.000000       7.000000   
50%         2.000000       9.000000       8.000000       9.000000   
75%         3.000000      11.000000      10.000000      11.000000   
max        20.000000      20.000000      20.000000      20.000000   

                 Agi            Agg            Aer            Acc  
count  144750.000000  144750.000000  144750.000000  144750.000000  
mean       11.097133      10.113202       2.996560      11.669941  
std         2.185484       3.308855       2.871859       2.166215  
min         1.000000       1.000000       1.000000       1.000000  
25%        10.000000       8.000000       1.000000      10.000000  
50%        11.000000      10.000000       2.000000      12.000000  
75%        13.000000      13.000000       3.000000      13.000000  
max        20.000000      20.000000      20.000000      20.000000  

[8 rows x 55 columns]

#### Observation:

* There is no odd values based on range of each column
* almost all numerical columns range from 1 to 20, except height, weight, age, value, wage, ca and pa
* if convert 'int64' to 'int16', we must skip value and wage columns! 

### Value Counts for Categorical Columns

In [5]:
df['Division'].value_counts()

Unknown                                         21110
Argentine Premier Division                       1758
Italian Serie A                                  1735
Sky Bet Championship                             1488
Brazilian National Lower Division                1452
                                                ...  
Swedish Third Division North Svealand               1
Swedish Third Division Middle Götaland             1
Romanian Fourth League Constanţa                   1
Moldovan Divizia A                                  1
Spanish Regional Preferente Extremadura Gr.1        1
Name: Division, Length: 1216, dtype: int64

#### Observation:

There are so many players from an unknown division. Since we have large dataset, we could drop this so we only process the active players from known divisions.

In [6]:
#remove players from 'unknown' divisions
df = df[df['Division']!='Unknown']
df.shape

(123640, 64)

In [7]:
df['Nation'].value_counts()

ARG    8203
BRA    6462
FRA    4840
ENG    4386
ESP    4292
       ... 
DJI       1
TCA       1
DMA       1
MON       1
LCA       1
Name: Nation, Length: 212, dtype: int64

In [8]:
df['Preferred Foot'].value_counts()

Right         59856
Right Only    35115
Left          15131
Left Only     10518
Either         3020
Name: Preferred Foot, dtype: int64

#### Observation:

Let's simplify the unique value into Right, Left, and Either

In [9]:
#simplify the preferred foot values
df['Preferred Foot Simple'] = df.apply(lambda x: x['Preferred Foot'].replace('Only', '').strip(), axis=1)
df.groupby(['Preferred Foot Simple', 'Preferred Foot']).size()

Preferred Foot Simple  Preferred Foot
Either                 Either             3020
Left                   Left              15131
                       Left Only         10518
Right                  Right             59856
                       Right Only        35115
dtype: int64

In [10]:
df['Best Pos'].value_counts()

D (C)     19261
ST (C)    18876
M (C)     17108
GK        13124
D (R)      9644
AM (R)     8359
D (L)      8301
DM         7430
AM (C)     6814
AM (L)     6807
M (R)      3805
M (L)      3332
WB (L)      504
WB (R)      275
Name: Best Pos, dtype: int64

#### Observation:

We will use this column instead of 'Position' column due to its value uniqueness. To simplify the category of best position, we will create new column based on the rule:
* GK
* D(C)/D(R)/D(L) --> D
* WB(R)/WB(L) --> WB
* DM
* M(C)/M(R)/M(L) --> M
* AM(C)/AM(R)/AM(L) --> AM
* ST(C) --> ST

In [11]:
#simplify best pos values
import re
df['Best Pos Simple'] = df.apply(lambda x: re.sub("\(.*?\)", "", x['Best Pos'].replace(' ', '')), axis=1)

df.groupby(['Best Pos Simple', 'Best Pos']).size()

Best Pos Simple  Best Pos
AM               AM (C)       6814
                 AM (L)       6807
                 AM (R)       8359
D                D (C)       19261
                 D (L)        8301
                 D (R)        9644
DM               DM           7430
GK               GK          13124
M                M (C)       17108
                 M (L)        3332
                 M (R)        3805
ST               ST (C)      18876
WB               WB (L)        504
                 WB (R)        275
dtype: int64

In [12]:
df['Best Role'].value_counts()

W          15753
CD         15190
FB         11878
P           9338
CM          9017
SK          6814
G           6293
IW          5270
AP          4345
BWM         4190
WB          3972
AF          3909
NCB         3171
DLP         2497
PF          2410
TM          2288
MEZ         2285
AM          2060
NFB         1758
SS          1345
CAR         1315
A           1309
BBM          970
DM           900
DW           871
BPD          837
CWB          551
IWB          535
VOL          447
F9           436
DLF          379
HB           320
EG           227
Unknown      184
IF           137
WTM           92
T             90
RPM           61
RGA           57
CF            49
WP            37
L             31
RMD           16
WM             6
Name: Best Role, dtype: int64

### Data Preprocessing

In [13]:
#drop first column
df = df.drop(columns = df.columns[0])

In [14]:
#change data type 'int64' to 'int16'
for column in df.columns:
    if df[column].dtypes == 'int64' and not (column in ['Value', 'Wage']):
        df[column] = df[column].astype('int16')

In [15]:
#remove whitespace
df.columns = df.columns.str.replace(' ', '').str.lower()

In [16]:
#collect columns based on data type
categorical = []
numerical = []
for column in df.columns:
    if df[column].dtypes == 'int16':
        numerical.append(column)
    else:
        categorical.append(column)

In [17]:
#collect columns based on atribute categories
profile = ['name', 'position', 'club', 'division', 'based', 'nation', 'age', 
           'preferredfoot', 'preferredfootsimple', 'bestpos', 'bestpossimple', 'bestrole', 'value', 'wage', 'ca', 'pa']
mental = ['wor', 'vis', 'tea', 'pos', 'otb', 'ldr', 'fla', 'det', 'dec', 'cnt', 'cmp', 'bra', 'ant', 'agg']
technical = ['tec', 'tck', 'pen', 'pas', '1v1', 'mar', 'lth', 'lon', 'hea', 'fre', 'fir', 'fin', 'dri', 'cro', 'cor']
physical = ['str', 'sta', 'pac', 'nat', 'jum', 'bal', 'agi', 'aer', 'acc', 'height', 'weight']
goalkeeping = ['thr', 'tro', 'ref', 'pun', 'kic', 'han', 'ecc', 'com', 'cmd']

mental = sorted(mental)
technical = sorted(technical)
physical = sorted(physical)
goalkeeping = sorted(goalkeeping)

In [18]:
#reorder columns based on attributes
order = profile + technical + mental + physical + goalkeeping
df = df[order]
df.head()

name                position       club  \
0       Lionel Messi         AM (RC), ST (C)  Barcelona   
1  Cristiano Ronaldo         AM (RL), ST (C)   Juventus   
2     Kylian Mbappé         AM (RL), ST (C)   Paris SG   
3       Manuel Neuer                      GK  FC Bayern   
4             Neymar  M (L), AM (LC), ST (C)   Paris SG   

                 division                       based nation  age  \
0  Spanish First Division      Spain (First Division)    ARG   32   
1         Italian Serie A             Italy (Serie A)    POR   34   
2       Ligue 1 Conforama  France (Ligue 1 Conforama)    FRA   20   
3              Bundesliga        Germany (Bundesliga)    GER   33   
4       Ligue 1 Conforama  France (Ligue 1 Conforama)    BRA   27   

  preferredfoot preferredfootsimple bestpos  ... weight cmd  com  ecc  han  \
0          Left                Left  AM (R)  ...     72   2    2    2    2   
1        Either              Either  ST (C)  ...     83   1    4    1    3   
2         Right               Right  ST (C)  ...     73   1    2    3    3   
3        Either              Either      GK  ...     90  17   12   18   14   
4         Right               Right  AM (L)  ...     68   1    3    1    2   

   kic  pun  ref  thr  tro  
0    1    2    1    2    3  
1    3    2    2    2    3  
2    4    1    1    1    2  
3   18    8   14   19   20  
4    2    2    4    1    1  

[5 rows x 65 columns]

In [19]:
#check duplicated data
df.duplicated().sum()

0

In [20]:
#check the last updated columns and orders
df.columns

Index(['name', 'position', 'club', 'division', 'based', 'nation', 'age',
       'preferredfoot', 'preferredfootsimple', 'bestpos', 'bestpossimple',
       'bestrole', 'value', 'wage', 'ca', 'pa', '1v1', 'cor', 'cro', 'dri',
       'fin', 'fir', 'fre', 'hea', 'lon', 'lth', 'mar', 'pas', 'pen', 'tck',
       'tec', 'agg', 'ant', 'bra', 'cmp', 'cnt', 'dec', 'det', 'fla', 'ldr',
       'otb', 'pos', 'tea', 'vis', 'wor', 'acc', 'aer', 'agi', 'bal', 'height',
       'jum', 'nat', 'pac', 'sta', 'str', 'weight', 'cmd', 'com', 'ecc', 'han',
       'kic', 'pun', 'ref', 'thr', 'tro'],
      dtype='object')

### Sampling

This dataset is huge. It still contains 123640 rows even after dropping some rows. So we need sampling for faster analysis and machine learning process. First, we must decide how many rows for sample by slovin formula.

n = N / (1 + Ne2)

Where:

n = Number of samples,
N = Total population and
e = Error tolerance (level).

In [24]:
#calculate sample size with slovin formula with 3% error tolerance
N = df.shape[0]
e = 0.03

sample_size = N/ (1 + N*(e**2))
sample_size

1101.2148633724037

#### Top 5 division with the highest value

In [25]:
#determine top 5 division
top5division = df.groupby('division').agg({'value':'mean', 'name':'count'}).sort_values(by='value', ascending=False).reset_index().head(5)
top5division

division         value  name
0  English Premier Division  8.266255e+06  1099
1    Spanish First Division  7.863478e+06   732
2                Bundesliga  5.574832e+06   674
3           Italian Serie A  2.838155e+06  1735
4         Ligue 1 Conforama  2.682701e+06   877

### Create Sample Dataset

In [26]:
#create sample
dfsample = df[df['division'] == 'English Premier Division']
dfsample.shape

(1099, 65)

based on our previous slovin sample size calculation, with this number (1099) we would have 3% error tolerance (sample size = 1101) by using English Premier Division players.

### Save Players Subset from English Premier Division to CSV File

In [27]:
dfsample.to_csv('englishpremierdivision2020.csv')